### PR TITLE
Predict validation video

### PR DESCRIPTION
**Pull request by annikabrundyn**

### CODE REVIEW

File: src/predict/predict.py [hunk 1]
```diff
@@ -1,45 +1,62 @@
...
+        # if (idx+1) % 2 == 0:
+        #     outputs = torch.cat(outputs)
+        #     torchvision.io.write_video(filename=os.path.join(args.save_path, f"{idx}.avi"),
+        #                                video_array=outputs,
+        #                                fps=30)
+        #     outputs = []
 
-test_results = trainer.test(model, test_dataloaders=dm.vis_img_dataloader())
+    outputs = torch.cat(outputs)
+    torchvision.io.write_video(filename=os.path.join(args.save_path, f"{idx}.avi"), video_array=outputs, fps=1)
```
Comment (jeswan) [inline src/predict/predict.py]: Let's set fps as an argument. It is likely in the 15-30 range (we'll likely have to approximate by eye).

File: src/predict/predict2.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        left_and_right = torch.clamp(left_and_right, 0, 1) * 255
+        outputs.append(left_and_right)
+
+        # output in chunks to avoid memory errors
+        if (idx + 1) % 5 == 0:
+            outputs = torch.cat(outputs).cpu()
+            torchvision.io.write_video(filename=os.path.join(args.output_dir, f"{idx}.avi"), video_array=outputs, fps=30)
+            outputs = []
+
+        outputs = torch.cat(outputs).cpu()
```
Comment (jeswan) [inline src/predict/predict2.py]: let's use gpu when available.

File: src/predict/predict2.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        torchvision.io.write_video(filename=os.path.join(args.output_dir, f"{idx}.avi"), video_array=outputs, fps=1)
```
Comment (jeswan) [inline src/predict/predict2.py]: frame rate should be same as above

File: src/predict/predict2.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+        # concat left view and right view
+        left_and_right = torch.cat((img[:, 0, ...], preds), dim=3).permute(0,2,3,1)
+        left_and_right = torch.clamp(left_and_right, 0, 1) * 255
+        outputs.append(left_and_right)
+
+        # output in chunks to avoid memory errors
+        if (idx + 1) % 5 == 0:
+            outputs = torch.cat(outputs).cpu()
+            torchvision.io.write_video(filename=os.path.join(args.output_dir, f"{idx}.avi"), video_array=outputs, fps=30)
```
Comment (jeswan) [inline src/predict/predict2.py]: what video player are you using to open these videos?

File: src/predict/predict2.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        batch_size=model.hparams.batch_size,
+        num_workers=model.hparams.num_workers,
+    )
+    dm.setup()
+    print("dm setup")
+
+    outputs = []
+    for idx, batch in enumerate(tqdm(dm.val_dataloader())):
+        if idx > 1:
+            break
```
Comment (jeswan) [inline src/predict/predict2.py]: why is this necessary? or is it just a remnant from testing?

File: src/data/datamodule.py [hunk 6]
```diff
@@ -120,6 +120,9 @@ def setup(self):
         # create separate list of validation images to visualize (hand-picked 14 indices)
         self.vis_samples = self._create_val_img_list()
 
+        # reverse to be from first to last frame for predicting videos since we don't shuffle
```
Comment (annikabrundyn) [inline src/data/datamodule.py:123]: @jeswan this shouldnt affect anything but now just making validation samples go forwards in time instead of backwards

File: src/data/dataset.py [hunk 7]
```diff
@@ -70,9 +70,7 @@ def __getitem__(self, index):
         target = self.target_transform(target)
 
         if self.extra_info:
-            sample_info = {}
-            sample_info['image_set'] = image_set
-            sample_info['frame_nums'] = " ".join(frames)
-            return image_tensor, target, sample_info
+            target_frame_name = str(frames[0])
```
Comment (annikabrundyn) [inline src/data/dataset.py:73]: extra_info is set to false everywhere so this wont affect anything either

File: src/models/right_view/combine_fns.py [hunk 8]
```diff
@@ -1,5 +1,5 @@
 import torch.nn as nn
-from convlstm import ConvLSTM
+from .convlstm import ConvLSTM
```
Comment (annikabrundyn) [inline src/models/right_view/combine_fns.py:2]: needed to fix this import

File: src/predict/predict_stacked.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
+import numpy as np
```
Comment (annikabrundyn) [inline src/predict/predict_stacked.py:1]: stacked vs unstacked should be one file with an argument but much easier to do separately for now - maybe we can change this in future! leaving like this for now

**Comment by annikabrundyn** [comment]
@jeswan i've check that this doesn't affect anything else - basically just adding 2 scripts to load checkpoint and predict video. can we merge so i can close the branch?
